Add batched grad testing to gradcheck, turn it on in test_autograd

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49120 Add batched grad testing to gradcheck, turn it on in test_autograd**

This adds a `check_batched_grad=False` option to gradcheck and gradgradcheck.
It defaults to False because gradcheck is a public API and I don't want
to break any existing non-pytorch users of gradcheck.
This:
- runs grad twice with two grad outputs, a & b
- runs a vmapped grad with torch.stack([a, b])
- compares the results of the above against each other.

Furthermore:
- `check_batched_grad=True` is set to be the default for
gradcheck/gradgradcheck inside of test_autograd.py. This is done by
reassigning to the gradcheck object inside test_autograd
- I manually added `check_batched_grad=False` to gradcheck instances
that don't support batched grad.
- I added a denylist for operations that don't support batched grad.

Question:
- Should we have a testing only gradcheck (e.g.,
torch.testing.gradcheck) that has different defaults from our public
API, torch.autograd.gradcheck?

Future:
- The future plan for this is to repeat the above for test_nn.py (the
autogenerated test will require a denylist)
- Finally, we can repeat the above for all pytorch test files that use
gradcheck.

Test Plan:
- run tests

Differential Revision: [D25563542](https://our.internmc.facebook.com/intern/diff/D25563542)

zou3519 committed:
Add batched grad testing to gradcheck, turn it on in test_autograd

This adds a `check_batched_grad=False` option to gradcheck and gradgradcheck.
It defaults to False because gradcheck is a public API and I don't want
to break any existing non-pytorch users of gradcheck.
This:
- runs grad twice with two grad outputs, a & b
- runs a vmapped grad with torch.stack([a, b])
- compares the results of the above against each other.

Furthermore:
- `check_batched_grad=True` is set to be the default for
gradcheck/gradgradcheck inside of test_autograd.py. This is done by
reassigning to the gradcheck object inside test_autograd
- I manually added `check_batched_grad=False` to gradcheck instances
that don't support batched grad.
- I added a denylist for operations that don't support batched grad.

Question:
- Should we have a testing only gradcheck (e.g.,
torch.testing.gradcheck) that has different defaults from our public
API, torch.autograd.gradcheck?

Future:
- The future plan for this is to repeat the above for test_nn.py (the
autogenerated test will require a denylist)
- Finally, we can repeat the above for all pytorch test files that use
gradcheck.

Test Plan:
- run tests

[ghstack-poisoned]

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -201,6 +203,32 @@ def get_analytical_jacobian(input, output, nondet_tol=0.0, grad_out=1.0):
+    for res, exp in zip(result, expected):
+        if torch.allclose(res, exp):
+            continue
+        return fail_test("TODO")
```
TODO: a nice error message telling folks what they should do if they see a test failure.

I'm trying to get signal from this PR first though on the CI

zou3519 committed:
Update on "Add batched grad testing to gradcheck, turn it on in test_autograd"


This adds a `check_batched_grad=False` option to gradcheck and gradgradcheck.
It defaults to False because gradcheck is a public API and I don't want
to break any existing non-pytorch users of gradcheck.
This:
- runs grad twice with two grad outputs, a & b
- runs a vmapped grad with torch.stack([a, b])
- compares the results of the above against each other.

Furthermore:
- `check_batched_grad=True` is set to be the default for
gradcheck/gradgradcheck inside of test_autograd.py. This is done by
reassigning to the gradcheck object inside test_autograd
- I manually added `check_batched_grad=False` to gradcheck instances
that don't support batched grad.
- I added a denylist for operations that don't support batched grad.

Question:
- Should we have a testing only gradcheck (e.g.,
torch.testing.gradcheck) that has different defaults from our public
API, torch.autograd.gradcheck?

Future:
- The future plan for this is to repeat the above for test_nn.py (the
autogenerated test will require a denylist)
- Finally, we can repeat the above for all pytorch test files that use
gradcheck.

Test Plan:
- run tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add batched grad testing to gradcheck, turn it on in test_autograd"


This adds a `check_batched_grad=False` option to gradcheck and gradgradcheck.
It defaults to False because gradcheck is a public API and I don't want
to break any existing non-pytorch users of gradcheck.
This:
- runs grad twice with two grad outputs, a & b
- runs a vmapped grad with torch.stack([a, b])
- compares the results of the above against each other.

Furthermore:
- `check_batched_grad=True` is set to be the default for
gradcheck/gradgradcheck inside of test_autograd.py. This is done by
reassigning to the gradcheck object inside test_autograd
- I manually added `check_batched_grad=False` to gradcheck instances
that don't support batched grad.
- I added a denylist for operations that don't support batched grad.

Question:
- Should we have a testing only gradcheck (e.g.,
torch.testing.gradcheck) that has different defaults from our public
API, torch.autograd.gradcheck?

Future:
- The future plan for this is to repeat the above for test_nn.py (the
autogenerated test will require a denylist)
- Finally, we can repeat the above for all pytorch test files that use
gradcheck.

Test Plan:
- run tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add batched grad testing to gradcheck, turn it on in test_autograd"


This adds a `check_batched_grad=False` option to gradcheck and gradgradcheck.
It defaults to False because gradcheck is a public API and I don't want
to break any existing non-pytorch users of gradcheck.
This:
- runs grad twice with two grad outputs, a & b
- runs a vmapped grad with torch.stack([a, b])
- compares the results of the above against each other.

Furthermore:
- `check_batched_grad=True` is set to be the default for
gradcheck/gradgradcheck inside of test_autograd.py. This is done by
reassigning to the gradcheck object inside test_autograd
- I manually added `check_batched_grad=False` to gradcheck instances
that don't support batched grad.
- I added a denylist for operations that don't support batched grad.

Question:
- Should we have a testing only gradcheck (e.g.,
torch.testing.gradcheck) that has different defaults from our public
API, torch.autograd.gradcheck?

Future:
- The future plan for this is to repeat the above for test_nn.py (the
autogenerated test will require a denylist)
- Finally, we can repeat the above for all pytorch test files that use
gradcheck.

Test Plan:
- run tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add batched grad testing to gradcheck, turn it on in test_autograd"


This adds a `check_batched_grad=False` option to gradcheck and gradgradcheck.
It defaults to False because gradcheck is a public API and I don't want
to break any existing non-pytorch users of gradcheck.
This:
- runs grad twice with two grad outputs, a & b
- runs a vmapped grad with torch.stack([a, b])
- compares the results of the above against each other.

Furthermore:
- `check_batched_grad=True` is set to be the default for
gradcheck/gradgradcheck inside of test_autograd.py. This is done by
reassigning to the gradcheck object inside test_autograd
- I manually added `check_batched_grad=False` to gradcheck instances
that don't support batched grad.
- I added a denylist for operations that don't support batched grad.

Question:
- Should we have a testing only gradcheck (e.g.,
torch.testing.gradcheck) that has different defaults from our public
API, torch.autograd.gradcheck?

Future:
- The future plan for this is to repeat the above for test_nn.py (the
autogenerated test will require a denylist)
- Finally, we can repeat the above for all pytorch test files that use
gradcheck.

Test Plan:
- run tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add batched grad testing to gradcheck, turn it on in test_autograd"


This adds a `check_batched_grad=False` option to gradcheck and gradgradcheck.
It defaults to False because gradcheck is a public API and I don't want
to break any existing non-pytorch users of gradcheck.
This:
- runs grad twice with two grad outputs, a & b
- runs a vmapped grad with torch.stack([a, b])
- compares the results of the above against each other.

Furthermore:
- `check_batched_grad=True` is set to be the default for
gradcheck/gradgradcheck inside of test_autograd.py. This is done by
reassigning to the gradcheck object inside test_autograd
- I manually added `check_batched_grad=False` to gradcheck instances
that don't support batched grad.
- I added a denylist for operations that don't support batched grad.

Question:
- Should we have a testing only gradcheck (e.g.,
torch.testing.gradcheck) that has different defaults from our public
API, torch.autograd.gradcheck?

Future:
- The future plan for this is to repeat the above for test_nn.py (the
autogenerated test will require a denylist)
- Finally, we can repeat the above for all pytorch test files that use
gradcheck.

Test Plan:
- run tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add batched grad testing to gradcheck, turn it on in test_autograd"


This adds a `check_batched_grad=False` option to gradcheck and gradgradcheck.
It defaults to False because gradcheck is a public API and I don't want
to break any existing non-pytorch users of gradcheck.
This:
- runs grad twice with two grad outputs, a & b
- runs a vmapped grad with torch.stack([a, b])
- compares the results of the above against each other.

Furthermore:
- `check_batched_grad=True` is set to be the default for
gradcheck/gradgradcheck inside of test_autograd.py. This is done by
reassigning to the gradcheck object inside test_autograd
- I manually added `check_batched_grad=False` to gradcheck instances
that don't support batched grad.
- I added a denylist for operations that don't support batched grad.

Question:
- Should we have a testing only gradcheck (e.g.,
torch.testing.gradcheck) that has different defaults from our public
API, torch.autograd.gradcheck?

Future:
- The future plan for this is to repeat the above for test_nn.py (the
autogenerated test will require a denylist)
- Finally, we can repeat the above for all pytorch test files that use
gradcheck.

Test Plan:
- run tests

[ghstack-poisoned]

ezyang commented:
Added @mruberry as this is touching `common_method_invocations`

ezyang commented on torch/autograd/gradcheck.py:
```diff
@@ -201,6 +203,59 @@ def get_analytical_jacobian(input, output, nondet_tol=0.0, grad_out=1.0):
+    grad_outputs = [torch.randn_like(output) for _ in range(2)]
+
+    expected = [vjp(gO) for gO in grad_outputs]
+    expected = [torch.stack(shards) for shards in zip(*expected)]
```
This isn't really a gradcheck right? A gradcheck is defined by comparing the analytic and numeric result. What you're doing here is verifying the functoriality of vmap over vjap.

ezyang replied:
(There's nothing wrong with this test, but not sure if it makes sense as an add on to gradcheck in this case.)

zou3519 replied:
That's correct. I can add a comment to this describing this.

One alternative to the approach here is to compute the full analytical jacobian via vmap. The problem I have with that is performance: it seems like it would do a bit more work than what is currently implemented in this PR. However, benchmarking shows that it's not that much more work: 191m (cpu time) vs 195m (cpu time, this PR) vs 198m (cpu time, if we computed the full jacobian)

ezyang replied:
I do agree that this testing strategy is better. My point is mostly an organizational one.

zou3519 replied:
In an ideal world I would separate out the "batched grad" check from gradcheck and have separate tests for them. In PyTorch today this is a little difficult without writing a lot of code: we use gradcheck all over the place, but our only "structured" testing happens in test_nn (in the ModuleTests) and test_autograd (with common_method_invocations).
These two sources (test_nn and the common_method_invocations table, and soon to be OpInfo) are not comprehensive and we end up manually invoking `gradcheck` across our tests.

So an alternative could be to put the batched grad testing into its own helper function `check_vmap_over_vjp`, hunt down all of the callsites of gradcheck, and call `check_vmap_over_vjp` right after said gradchecks. There's around 430 places where this happens which will take me a while to hunt down, but on the upside it gives the right organizational structure.

zou3519 replied:
I think for now keeping the organization this way makes it so that it is easier to test things; I can see splitting the functionality up as a way to implement OpInfo based tests. I'll add a comment into the code saying something about this.

ezyang commented on test/test_autograd.py:
```diff
@@ -3871,14 +3874,14 @@ def backward(ctx, grad_out):
 
         inp = torch.randn(5, 5, requires_grad=True)
-        gradcheck(lambda x: NonDetFunc.apply(x, 0.0), inp)
+        gradcheck(lambda x: NonDetFunc.apply(x, 0.0), inp, check_batched_grad=False)
```
Huh, didn't know this works.

albanD replied:
This works in what sense? That we have a proper test for non-determinism in the backward?

mruberry commented on test/test_autograd.py:
```diff
@@ -4877,18 +4880,22 @@ def gradgradcheck_method_precision_override(test_name):
 def run_grad_and_gradgrad_checks(test_case, name, test_name, apply_method, output_variable,
-                                 input_variables, run_gradgradcheck=True):
-    test_case.assertTrue(gradcheck(apply_method, input_variables, eps=1e-6, atol=PRECISION))
+                                 input_variables, run_gradgradcheck=True, check_batched_grad=True):
```
Tests using method_tests() are being deprecated in favor of tests using OpInfos like these:

https://github.com/pytorch/pytorch/blob/bd322c8967b99551df67d479ebcc435c030980c9/test/test_ops.py#L57

I'm developing a post on this now, but the plan is to remove method_tests(). This PR doesn't have to update the OpInfo-based testing but that should happen in the near future to continue testing this property. Instead of a custom `EXCLUDE_BATCHED_GRAD_TESTS` a property could be added to the OpInfo class for whether to run batched grad tests or not, or the OpInfo skip mechanism can be reused.

zou3519 replied:
I'm happy to add something to the OpInfo class in a followup to make sure these are tested in the future.

ezyang commented on test/test_autograd.py:
```diff
@@ -71,6 +71,9 @@ def getattr_qualified(obj, qname, default=None):
 PRECISION = 1e-4
 
+gradcheck = partial(gradcheck, check_batched_grad=True)
+gradgradcheck = partial(gradgradcheck, check_batched_grad=True)
```
I was going to complain about how this is very unexpected but then I went and looked to see how many occurrences of gradcheck there are in this file and... well... there are a lot.

ezyang commented:
Going to give mruberry time to look before approving or not. This basically LGTM though.

mruberry commented:
> Going to give mruberry time to look before approving or not. This basically LGTM though.

No worries on updating this test, it is still heavily used at the moment. But the OpInfo-based tests will need to be updated, too. It's OK if that update happens in a future PR.

zou3519 commented:
> > Going to give mruberry time to look before approving or not. This basically LGTM though.
> 
> No worries on updating this test, it is still heavily used at the moment. But the OpInfo-based tests will need to be updated, too. It's OK if that update happens in a future PR.

I'm happy to update the OpInfo tests in a future PR. I'm also planning on turning this check on in test_nn.py as well in a followup -- is there something similar to OpInfo coming for test_nn soon that I also update?

mruberry commented:
> > > Going to give mruberry time to look before approving or not. This basically LGTM though.
> > 
> > 
> > No worries on updating this test, it is still heavily used at the moment. But the OpInfo-based tests will need to be updated, too. It's OK if that update happens in a future PR.
> 
> @mruberry I'm happy to update the OpInfo tests in a future PR. I'm also planning on turning this check on in test_nn.py as well in a followup -- is there something similar to OpInfo coming for test_nn soon that I also update?

Maybe. Functions in torch.nn.functional can use the OpInfo pattern we have, but Modules may need a new ModuleInfo pattern. There's been interest in developing that in H1 but we don't have a concrete plan yet.

zou3519 committed:
Update on "Add batched grad testing to gradcheck, turn it on in test_autograd"


This adds a `check_batched_grad=False` option to gradcheck and gradgradcheck.
It defaults to False because gradcheck is a public API and I don't want
to break any existing non-pytorch users of gradcheck.
This:
- runs grad twice with two grad outputs, a & b
- runs a vmapped grad with torch.stack([a, b])
- compares the results of the above against each other.

Furthermore:
- `check_batched_grad=True` is set to be the default for
gradcheck/gradgradcheck inside of test_autograd.py. This is done by
reassigning to the gradcheck object inside test_autograd
- I manually added `check_batched_grad=False` to gradcheck instances
that don't support batched grad.
- I added a denylist for operations that don't support batched grad.

Question:
- Should we have a testing only gradcheck (e.g.,
torch.testing.gradcheck) that has different defaults from our public
API, torch.autograd.gradcheck?

Future:
- The future plan for this is to repeat the above for test_nn.py (the
autogenerated test will require a denylist)
- Finally, we can repeat the above for all pytorch test files that use
gradcheck.

Test Plan:
- run tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add batched grad testing to gradcheck, turn it on in test_autograd"


This adds a `check_batched_grad=False` option to gradcheck and gradgradcheck.
It defaults to False because gradcheck is a public API and I don't want
to break any existing non-pytorch users of gradcheck.
This:
- runs grad twice with two grad outputs, a & b
- runs a vmapped grad with torch.stack([a, b])
- compares the results of the above against each other.

Furthermore:
- `check_batched_grad=True` is set to be the default for
gradcheck/gradgradcheck inside of test_autograd.py. This is done by
reassigning to the gradcheck object inside test_autograd
- I manually added `check_batched_grad=False` to gradcheck instances
that don't support batched grad.
- I added a denylist for operations that don't support batched grad.

Question:
- Should we have a testing only gradcheck (e.g.,
torch.testing.gradcheck) that has different defaults from our public
API, torch.autograd.gradcheck?

Future:
- The future plan for this is to repeat the above for test_nn.py (the
autogenerated test will require a denylist)
- Finally, we can repeat the above for all pytorch test files that use
gradcheck.

Test Plan:
- run tests

[ghstack-poisoned]

mruberry commented:
I'll let @albanD and @ezyang comment on the code. My question is: how much do we expect test time to change with this diff?

zou3519 commented:
@mruberry, not very much. On my (very noisy) machine:
```
Before:
real    8m3.991s
user    191m13.005s
sys     0m59.089s

After:
real    8m18.060s
user    195m27.363s
sys     1m5.554s
```

mruberry commented:
> @mruberry, not very much. On my (very noisy) machine:
> 
> ```
> Before:
> real    8m3.991s
> user    191m13.005s
> sys     0m59.089s
> 
> After:
> real    8m18.060s
> user    195m27.363s
> sys     1m5.554s
> ```

I also took a look at the timings on the CI machines and it seems like a minimal difference.

zou3519 committed:
Update on "Add batched grad testing to gradcheck, turn it on in test_autograd"


This adds a `check_batched_grad=False` option to gradcheck and gradgradcheck.
It defaults to False because gradcheck is a public API and I don't want
to break any existing non-pytorch users of gradcheck.
This:
- runs grad twice with two grad outputs, a & b
- runs a vmapped grad with torch.stack([a, b])
- compares the results of the above against each other.

Furthermore:
- `check_batched_grad=True` is set to be the default for
gradcheck/gradgradcheck inside of test_autograd.py. This is done by
reassigning to the gradcheck object inside test_autograd
- I manually added `check_batched_grad=False` to gradcheck instances
that don't support batched grad.
- I added a denylist for operations that don't support batched grad.

Question:
- Should we have a testing only gradcheck (e.g.,
torch.testing.gradcheck) that has different defaults from our public
API, torch.autograd.gradcheck?

Future:
- The future plan for this is to repeat the above for test_nn.py (the
autogenerated test will require a denylist)
- Finally, we can repeat the above for all pytorch test files that use
gradcheck.

Test Plan:
- run tests

Differential Revision: [D25563542](https://our.internmc.facebook.com/intern/diff/D25563542)

[ghstack-poisoned]

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -401,6 +457,15 @@ def not_reentrant_error(error_str=''):
             return not_reentrant_error(' (calculated using complex valued grad output)')
 
+        if check_batched_grad:
+            assert reentrant
```
This will make gradcheck raise an error even when raise_exception=False
Should this be handled via the `fail_test` function like all other failures here?

zou3519 replied:
This assertion is intentional, the code to check batched grad assumes that backward is reentrant. If someone e.g. moved the batched grad check to before failing with the reentrant check, this will give a loud error.

zou3519 replied:
I'll add a comment, please let me know if that sounds reasonable

albanD replied:
I didn't mean to remove it, but just do `fail_test("vmap check is expected to fail when the backward is not reentrant, you should either fix the reentrant test or disable the vmap check if the backward is expected not to be reentrant")`.
The gradcheck function should never do a hard crash when `raise_exception=False`.

zou3519 replied:
The assert can never be triggered by users or authors of PyTorch tests. That is because above this line, we check for `reentrant` and do an early return if `reentrant` is True. If the assert gets triggered, then that means a developer who is modifying `gradcheck` has done something wrong.

This means that although there is an `assert` line, the gradcheck function (as long as it does not get modified) actually never does a hard crash when `raise_exception=False`.

albanD replied:
Sounds good. I didn't knew that we are currently early exiting when not reentrant.
Could you just add a comment next to this assert then mentioning this?
Thanks for the explanation!

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -498,7 +563,8 @@ def gradgradcheck(
     check_undefined_grad: bool = True,
-    check_grad_dtypes: bool = False
+    check_grad_dtypes: bool = False,
+    check_batched_grad: bool = False,
```
Should the docstring be updated here?

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -234,7 +289,8 @@ def gradcheck(
     check_undefined_grad: bool = True,
-    check_grad_dtypes: bool = False
+    check_grad_dtypes: bool = False,
+    check_batched_grad: bool = False,
```
docstring update?

zou3519 replied:
Good catch, will update

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -201,6 +203,59 @@ def get_analytical_jacobian(input, output, nondet_tol=0.0, grad_out=1.0):
+    def vjp(v):
+        results = grad(v)
+        results = tuple(grad if grad is not None else
+                        torch.zeros([], dtype=output.dtype, device=output.device).expand(inp.shape)
```
nit: why not use input dtype and device if you use input shape here?

zou3519 committed:
Update on "Add batched grad testing to gradcheck, turn it on in test_autograd"


This adds a `check_batched_grad=False` option to gradcheck and gradgradcheck.
It defaults to False because gradcheck is a public API and I don't want
to break any existing non-pytorch users of gradcheck.
This:
- runs grad twice with two grad outputs, a & b
- runs a vmapped grad with torch.stack([a, b])
- compares the results of the above against each other.

Furthermore:
- `check_batched_grad=True` is set to be the default for
gradcheck/gradgradcheck inside of test_autograd.py. This is done by
reassigning to the gradcheck object inside test_autograd
- I manually added `check_batched_grad=False` to gradcheck instances
that don't support batched grad.
- I added a denylist for operations that don't support batched grad.

Question:
- Should we have a testing only gradcheck (e.g.,
torch.testing.gradcheck) that has different defaults from our public
API, torch.autograd.gradcheck?

Future:
- The future plan for this is to repeat the above for test_nn.py (the
autogenerated test will require a denylist)
- Finally, we can repeat the above for all pytorch test files that use
gradcheck.

Test Plan:
- run tests

Differential Revision: [D25563542](https://our.internmc.facebook.com/intern/diff/D25563542)

[ghstack-poisoned]

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -401,6 +459,17 @@ def not_reentrant_error(error_str=''):
             return not_reentrant_error(' (calculated using complex valued grad output)')
 
+        if check_batched_grad:
+            assert reentrant, ('Batched gradient checking makes the assumption that '
```
nit I would mention that we expect gradcheck to have early exited before reaching this point when not re-entrant.

albanD reviewed (approved):
LGTM thanks for the update!

zou3519 committed:
Update on "Add batched grad testing to gradcheck, turn it on in test_autograd"


This adds a `check_batched_grad=False` option to gradcheck and gradgradcheck.
It defaults to False because gradcheck is a public API and I don't want
to break any existing non-pytorch users of gradcheck.
This:
- runs grad twice with two grad outputs, a & b
- runs a vmapped grad with torch.stack([a, b])
- compares the results of the above against each other.

Furthermore:
- `check_batched_grad=True` is set to be the default for
gradcheck/gradgradcheck inside of test_autograd.py. This is done by
reassigning to the gradcheck object inside test_autograd
- I manually added `check_batched_grad=False` to gradcheck instances
that don't support batched grad.
- I added a denylist for operations that don't support batched grad.

Question:
- Should we have a testing only gradcheck (e.g.,
torch.testing.gradcheck) that has different defaults from our public
API, torch.autograd.gradcheck?

Future:
- The future plan for this is to repeat the above for test_nn.py (the
autogenerated test will require a denylist)
- Finally, we can repeat the above for all pytorch test files that use
gradcheck.

Test Plan:
- run tests

Differential Revision: [D25563542](https://our.internmc.facebook.com/intern/diff/D25563542)

[ghstack-poisoned]

zou3519 committed:
Update on "Add batched grad testing to gradcheck, turn it on in test_autograd"


This adds a `check_batched_grad=False` option to gradcheck and gradgradcheck.
It defaults to False because gradcheck is a public API and I don't want
to break any existing non-pytorch users of gradcheck.
This:
- runs grad twice with two grad outputs, a & b
- runs a vmapped grad with torch.stack([a, b])
- compares the results of the above against each other.

Furthermore:
- `check_batched_grad=True` is set to be the default for
gradcheck/gradgradcheck inside of test_autograd.py. This is done by
reassigning to the gradcheck object inside test_autograd
- I manually added `check_batched_grad=False` to gradcheck instances
that don't support batched grad.
- I added a denylist for operations that don't support batched grad.

Question:
- Should we have a testing only gradcheck (e.g.,
torch.testing.gradcheck) that has different defaults from our public
API, torch.autograd.gradcheck?

Future:
- The future plan for this is to repeat the above for test_nn.py (the
autogenerated test will require a denylist)
- Finally, we can repeat the above for all pytorch test files that use
gradcheck.

Test Plan:
- run tests

Differential Revision: [D25563542](https://our.internmc.facebook.com/intern/diff/D25563542)

[ghstack-poisoned]

malfet commented:
Reverting this change as it broke `TestAutograd.test_lobpcg`:
```
$ PYTORCH_TEST_WITH_SLOW=1 python test_autograd.py -v TestAutograd.test_lobpcg
test_lobpcg (__main__.TestAutograd) ... ERROR

======================================================================
ERROR: test_lobpcg (__main__.TestAutograd)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/nshulga/git/pytorch/torch/testing/_internal/common_utils.py", line 566, in wrapper
    fn(*args, **kwargs)
  File "/home/nshulga/git/pytorch/torch/testing/_internal/common_utils.py", line 526, in wrapper
    fn(*args, **kwargs)
  File "test_autograd.py", line 2851, in test_lobpcg
    run_symeig_test(1, (6, 6), largest=largest)
  File "test_autograd.py", line 2833, in run_symeig_test
    gradcheck(lambda A: func(k, A, largest), A)
  File "/home/nshulga/git/pytorch/torch/autograd/gradcheck.py", line 473, in gradcheck
    test_batched_grad(fail_test, tupled_inputs, o, j)
  File "/home/nshulga/git/pytorch/torch/autograd/gradcheck.py", line 253, in test_batched_grad
    result = vmap(vjp)(torch.stack(grad_outputs))
  File "/home/nshulga/git/pytorch/torch/_vmap_internals.py", line 250, in wrapped
    batched_outputs = func(*batched_inputs)
  File "/home/nshulga/git/pytorch/torch/autograd/gradcheck.py", line 238, in vjp
    results = grad(v)
  File "/home/nshulga/git/pytorch/torch/autograd/__init__.py", line 225, in grad
    inputs, allow_unused, accumulate_grad=False)
  File "/home/nshulga/git/pytorch/torch/autograd/function.py", line 89, in apply
    return self._forward_cls.backward(self, *args)  # type: ignore
  File "/home/nshulga/git/pytorch/torch/_lobpcg.py", line 328, in backward
    D_grad, U_grad, A, D, U, largest
  File "/home/nshulga/git/pytorch/torch/_lobpcg.py", line 259, in _symeig_backward
    D_grad, U_grad, A, D, U, largest
  File "/home/nshulga/git/pytorch/torch/_lobpcg.py", line 159, in _symeig_backward_partial_eigenspace
    generator=gen
RuntimeError: vmap: We do not yet support calling random operations inside of vmap. Please perform random operations outside of vmap as a workaround

----------------------------------------------------------------------
Ran 1 test in 0.440s

FAILED (errors=1)

```
